### PR TITLE
[Feature][Jdbc-Connector] Add OceanBase Connector

### PR DESCRIPTION
**Pull request by changhuyan**
## Purpose of this pull request

Add OceanBase Connector

## Check list

* [ ] Code changed are covered with tests, or it does not need tests for reason:
* [ ] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [ ] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs
* [ ] If you are contributing the connector code, please check that the following files are updated:
  1. Update change log that in connector document. For more details you can refer to [connector-v2]
Examples are as follows：
env {
    job.name = "SeaTunnel"
    spark.executor.instances = 1
    spark.executor.cores = 1
    spark.executor.memory = "1g"
    spark.master = local
}

source {
 jdbc{
      url = "jdbc:oceanbase://xxxxx:2881/test"
      driver = "com.alipay.oceanbase.jdbc.Driver"
      user = "xxx"
      password = "xxx"
      query = "select id,name from test"
      driver_type = "mysql"
 }
}

transform {
}

sink {
  jdbc {
       url = "jdbc:oceanbase://xxx:2881/test"
       driver = "com.alipay.oceanbase.jdbc.Driver"
       user = "root"
       password = "xxxxx"
       driver_type = "oracle"
       query = "insert into test3(id,name) values(?,?)"
  }
}
  3. Update [plugin-mapping.properties](https://github.com/apache/incubator-seatunnel/blob/dev/plugin-mapping.properties) and add new connector information in it
  4. Update the pom file of [seatunnel-dist](https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml)
* [ ] Update the [`release-note`](https://github.com/apache/incubator-seatunnel/blob/dev/release-note.md).

### CODE REVIEW

File: seatunnel-examples/seatunnel-engine-examples/src/main/java/org/apache/seatunnel/example/engine/SeaTunnelEngineExample.java [hunk 1]
```diff
@@ -31,7 +31,7 @@ public class SeaTunnelEngineExample {
 
     public static void main(String[] args)
             throws FileNotFoundException, URISyntaxException, CommandException {
-        String configurePath = args.length > 0 ? args[0] : "/examples/fake_to_console.conf";
+        String configurePath = args.length > 0 ? args[0] : "/examples/oceabase.conf";
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-engine-examples/src/main/java/org/apache/seatunnel/example/engine/SeaTunnelEngineExample.java]: revert

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/java/org/apache/seatunnel/example/spark/v2/SeaTunnelApiExample.java [hunk 2]
```diff
@@ -26,7 +26,7 @@ public class SeaTunnelApiExample {
 
     public static void main(String[] args)
             throws FileNotFoundException, URISyntaxException, CommandException {
-        String configurePath = args.length > 0 ? args[0] : "/examples/spark.batch.conf";
+        String configurePath = args.length > 0 ? args[0] : "/examples/oceabase.conf";
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/java/org/apache/seatunnel/example/spark/v2/SeaTunnelApiExample.java]: revert

**Review by hailin0 — COMMENTED** [review]
1. add e2e 
https://github.com/apache/incubator-seatunnel/tree/dev/seatunnel-e2e/seatunnel-connector-v2-e2e

2. add docs
https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/connector-v2/source/Jdbc.md
https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/connector-v2/sink/Jdbc.md

3. update `release-note.md`

**Comment by hailin0** [comment]
please check ci error
https://github.com/apache/incubator-seatunnel/actions/runs/4749964848/jobs/8438153776?pr=4626

**Comment by hailin0** [comment]
please check ci error

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/java/org/apache/seatunnel/example/spark/v2/SeaTunnelApiExample.java [hunk 3]
```diff
@@ -26,7 +26,7 @@ public class SeaTunnelApiExample {
 
     public static void main(String[] args)
             throws FileNotFoundException, URISyntaxException, CommandException {
-        String configurePath = args.length > 0 ? args[0] : "/examples/spark.batch.conf";
+        String configurePath = args.length > 0 ? args[0] : "/examples/fake_to_console.conf";
```
Comment (hailin0) [inline seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/java/org/apache/seatunnel/example/spark/v2/SeaTunnelApiExample.java:29]: revert

File: seatunnel-connectors-v2/connector-jdbc/pom.xml [hunk 4]
```diff
@@ -144,7 +151,10 @@
             <groupId>mysql</groupId>
             <artifactId>mysql-connector-java</artifactId>
         </dependency>
-
+        <dependency>
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/pom.xml:154]: add this drive into：

https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/pom.xml#L558
https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-dist/src/main/assembly/assembly-bin-ci.xml#L195

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/oceanbase/OceanBaseMySqlDialectFactory.java [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+import org.apache.seatunnel.connectors.seatunnel.jdbc.internal.dialect.JdbcDialectFactory;
+import org.apache.seatunnel.connectors.seatunnel.jdbc.internal.dialect.mysql.MysqlDialect;
+
+@AutoService(JdbcDialectFactory.class)
+public class OceanBaseMySqlDialectFactory implements JdbcDialectFactory {
+    @Override
+    public boolean acceptsURL(String url, Optional<String> driverTye) {
+        return url.startsWith("jdbc:oceanbase:")
+                && driverTye.isPresent()
+                && driverTye.get().equalsIgnoreCase("mysql");
```
Comment (hailin0) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/oceanbase/OceanBaseMySqlDialectFactory.java:34]: add example into docs

File: seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/java/org/apache/seatunnel/example/spark/v2/SeaTunnelApiExample.java (hunk 3 above)
Comment (changhuyan) [reply seatunnel-examples/seatunnel-spark-connector-v2-example/src/main/java/org/apache/seatunnel/example/spark/v2/SeaTunnelApiExample.java:29]: ok

File: seatunnel-connectors-v2/connector-jdbc/pom.xml (hunk 4 above)
Comment (changhuyan) [reply seatunnel-connectors-v2/connector-jdbc/pom.xml:154]: ok

**Comment by hailin0** [comment]
1. check ci
2. add e2e testcase

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+    private static final String DRIVER_CLASS = "com.alipay.oceanbase.jdbc.Driver";
+
+    private static final List<String> CONFIG_FILE =
+            Lists.newArrayList("/jdbc_oceanbase_mysql_source_and_oceanbase_oracle_sink.conf");
+
+    private static final String CREATE_SQL =
+            "CREATE TABLE `people` (\n"
+                    + "  `age` varchar(255)  DEFAULT NULL,\n"
+                    + "  `name` varchar(255) DEFAULT NULL\n"
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java:59]: Test all datatypes

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+                .sinkTable(OCEANBASE_SINK)
+                .createSql(CREATE_SQL)
+                .configFile(CONFIG_FILE)
+                .insertSql(insertSql)
+                .testData(testDataSet)
+                .build();
+    }
+
+    @Override
+    void compareResult() {}
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java:94]: implement this logic

File: docs/en/connector-v2/sink/Jdbc.md [hunk 8]
```diff
@@ -33,6 +33,7 @@ support `Xa transactions`. You can set `is_exactly_once=true` to enable it.
 | user                                      | String  | No       | -             |
 | password                                  | String  | No       | -             |
 | query                                     | String  | No       | -             |
+| driver_type                               | String  | No       | -             |
```
Comment (hailin0) [inline docs/en/connector-v2/sink/Jdbc.md:36]: add appendix & axample

https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/connector-v2/source/Jdbc.md#appendix

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+import org.testcontainers.utility.DockerLoggerFactory;
+
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Disabled("Disabled because it needs user's personal oceanbase account to run this test!")
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java:37]: Use docker deploy ob server

https://en.oceanbase.com/docs/community-observer-en-10000000000829647#Use%20Docker%20to%20deploy%20OceanBase%20Database

**Comment by NickCodeJourney** [comment]
why not use mysql connector

**Comment by changhuyan** [comment]
&nbsp;Because OceanBase databases are used in two ways,&nbsp;namely mysql and oracle.


------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/seatunnel"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年5月28日(星期天) 下午2:20
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/seatunnel] [Feature][Jdbc-Connector] Add OceanBase Connector (PR #4626)





 
why not use mysql connector

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/JdbcDialectFactory.java [hunk 10]
```diff
@@ -33,7 +35,7 @@ public interface JdbcDialectFactory {
      * @return <code>true</code> if this dialect understands the given URL; <code>false</code>
      *     otherwise.
      */
-    boolean acceptsURL(String url);
```
Comment (whhe) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/JdbcDialectFactory.java:36]: I think the `driverType` param shoud be passed in the create method below, and we can use a new default method to keep the other dialects unchanged.

```java
default JdbcDialect create(String driverType) {
    return create();
}
```

File: seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/oceanbase/OceanBaseMySqlDialectFactory.java [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    @Override
+    public JdbcDialect create() {
+        return new MysqlDialect();
```
Comment (whhe) [inline seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/oceanbase/OceanBaseMySqlDialectFactory.java:39]: We can merge the two dialect classes of OceanBase and return the JdbcDialect by the `driverType` param here.

```java
public JdbcDialect create(@Nonnull String driverType) {
    if ("mysql".equalsIgnoreCase(driverType)) {
        return new MysqlDialect();
    }
    return new OracleDialect();
}
```

File: seatunnel-connectors-v2/connector-jdbc/pom.xml [hunk 12]
```diff
@@ -129,6 +130,12 @@
                 <version>${vertica.version}</version>
                 <scope>provided</scope>
             </dependency>
+            <dependency>
+                <groupId>com.oceanbase</groupId>
+                <artifactId>oceanbase-client</artifactId>
+                <version>${oceanbase.version}</version>
+                <scope>provided</scope>
+            </dependency>
```
Comment (whhe) [inline seatunnel-connectors-v2/connector-jdbc/pom.xml:138]: It should not be imported directly due to the lgpl license https://github.com/oceanbase/obconnector-j

File: seatunnel-dist/pom.xml [hunk 13]
```diff
@@ -488,6 +489,12 @@
                 </dependency>
 
                 <!-- jdbc driver -->
+                <dependency>
+                    <groupId>com.alipay.oceanbase</groupId>
+                    <artifactId>oceanbase-client</artifactId>
+                    <version>${oceanbase.version}</version>
+                    <scope>provided</scope>
+                </dependency>
```
Comment (whhe) [inline seatunnel-dist/pom.xml:497]: Should not use the previous internal version, just keep it same with `seatunnel-connectors-v2/connector-jdbc/pom.xml`

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java [hunk 14]
```diff
@@ -0,0 +1,139 @@
...
+public class JdbcOceanbaseIT extends AbstractJdbcIT {
+    private static final String OCEANBASE_IMAGE = "shihd/oceanbase:1.0";
```
Comment (whhe) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java:39]: You can use the docker image `oceanbase/oceanbase-ce:4.0.0.0` with user `root` and empty password.

**Review by whhe — CHANGES_REQUESTED** [review]

**Comment by changhuyan** [comment]
Yes, there's a better way. You can change it.




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "apache/seatunnel"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年6月27日(星期二) 晚上8:04
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [apache/seatunnel] [Feature][Jdbc-Connector] Add OceanBase Connector (PR #4626)





 
@whhe requested changes on this pull request.
 
 
In seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/JdbcDialectFactory.java:
 &gt; @@ -33,7 +35,7 @@ public interface JdbcDialectFactory {       * @return <code&gt;true</code&gt; if this dialect understands the given URL; <code&gt;false</code&gt;       *     otherwise.       */ -    boolean acceptsURL(String url);  
I think the driverType param shoud be passed in the create method below, and we can use a new default method to keep the other dialects unchanged.
 default JdbcDialect create(String driverType) {     return create(); }
 
 
In seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/oceanbase/OceanBaseMySqlDialectFactory.java:
 &gt; +import com.google.auto.service.AutoService; + +import java.util.Optional; + ***@***.***(JdbcDialectFactory.class) +public class OceanBaseMySqlDialectFactory implements JdbcDialectFactory { +    @Override +    public boolean acceptsURL(String url, Optional<String&gt; driverTye) { +        return url.startsWith("jdbc:oceanbase:") +                &amp;&amp; driverTye.isPresent() +                &amp;&amp; driverTye.get().equalsIgnoreCase("mysql"); +    } + +    @Override +    public JdbcDialect create() { +        return new MysqlDialect();  
We can merge the two dialect classes of OceanBase and return the JdbcDialect by the driverType param here.
 public JdbcDialect ***@***.*** String driverType) {     if ("mysql".equalsIgnoreCase(driverType)) {         return new MysqlDialect();     }     return new OracleDialect(); }
 
 
In seatunnel-connectors-v2/connector-jdbc/pom.xml:
 &gt; @@ -129,6 +130,12 @@                  <version&gt;${vertica.version}</version&gt;                  <scope&gt;provided</scope&gt;              </dependency&gt; +            <dependency&gt; +                <groupId&gt;com.oceanbase</groupId&gt; +                <artifactId&gt;oceanbase-client</artifactId&gt; +                <version&gt;${oceanbase.version}</version&gt; +                <scope&gt;provided</scope&gt; +            </dependency&gt;  
It should not be imported directly due to the lgpl license https://github.com/oceanbase/obconnector-j
 
 
In seatunnel-dist/pom.xml:
 &gt; @@ -488,6 +489,12 @@                  </dependency&gt;                    <!-- jdbc driver --&gt; +                <dependency&gt; +                    <groupId&gt;com.alipay.oceanbase</groupId&gt; +                    <artifactId&gt;oceanbase-client</artifactId&gt; +                    <version&gt;${oceanbase.version}</version&gt; +                    <scope&gt;provided</scope&gt; +                </dependency&gt;  
Should not use the previous internal version, just keep it same with seatunnel-connectors-v2/connector-jdbc/pom.xml
 
 
In seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java:
 &gt; +import org.junit.jupiter.api.Disabled; +import org.testcontainers.containers.GenericContainer; +import org.testcontainers.containers.output.Slf4jLogConsumer; +import org.testcontainers.utility.DockerImageName; +import org.testcontainers.utility.DockerLoggerFactory; + +import com.google.common.collect.Lists; + +import java.util.ArrayList; +import java.util.HashMap; +import java.util.List; +import java.util.Map; + ***@***.***("Disabled because it needs user's personal oceanbase account to run this test!") +public class JdbcOceanbaseIT extends AbstractJdbcIT { +    private static final String OCEANBASE_IMAGE = "shihd/oceanbase:1.0";  
You can use the docker image oceanbase/oceanbase-ce:4.0.0.0 with user root and empty password.

**Comment by whhe** [comment]
@changhuyan Thanks for your contributation! I left some comments, and you can touch me at any time if you have any ideas on them.

For the reference to `oceanbase-client`, I think it may be an appropriate method to guide users to download it manually through the document, but I'm not sure how this affects the integration tests.

**Comment by whhe** [comment]
> Yes, there's a better way. You can change it.
> […](#)
> ------------------&nbsp;原始邮件&nbsp;------------------ 发件人: "apache/seatunnel" ***@***.***&gt;; 发送时间:&nbsp;2023年6月27日(星期二) 晚上8:04 ***@***.***&gt;; ***@***.******@***.***&gt;; 主题:&nbsp;Re: [apache/seatunnel] [Feature][Jdbc-Connector] Add OceanBase Connector (PR #4626) @whhe requested changes on this pull request. In seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/JdbcDialectFactory.java: &gt; @@ -33,7 +35,7 @@ public interface JdbcDialectFactory { * @return <code&gt;true</code&gt; if this dialect understands the given URL; <code&gt;false</code&gt; * otherwise. */ - boolean acceptsURL(String url); I think the driverType param shoud be passed in the create method below, and we can use a new default method to keep the other dialects unchanged. default JdbcDialect create(String driverType) { return create(); } In seatunnel-connectors-v2/connector-jdbc/src/main/java/org/apache/seatunnel/connectors/seatunnel/jdbc/internal/dialect/oceanbase/OceanBaseMySqlDialectFactory.java: &gt; +import com.google.auto.service.AutoService; + +import java.util.Optional; + ***@***.***(JdbcDialectFactory.class) +public class OceanBaseMySqlDialectFactory implements JdbcDialectFactory { + @OverRide + public boolean acceptsURL(String url, Optional<String&gt; driverTye) { + return url.startsWith("jdbc:oceanbase:") + &amp;&amp; driverTye.isPresent() + &amp;&amp; driverTye.get().equalsIgnoreCase("mysql"); + } + + @OverRide + public JdbcDialect create() { + return new MysqlDialect(); We can merge the two dialect classes of OceanBase and return the JdbcDialect by the driverType param here. public JdbcDialect ***@***.*** String driverType) { if ("mysql".equalsIgnoreCase(driverType)) { return new MysqlDialect(); } return new OracleDialect(); } In seatunnel-connectors-v2/connector-jdbc/pom.xml: &gt; @@ -129,6 +130,12 @@ <version&gt;${vertica.version}</version&gt; <scope&gt;provided</scope&gt; </dependency&gt; + <dependency&gt; + <groupId&gt;com.oceanbase</groupId&gt; + <artifactId&gt;oceanbase-client</artifactId&gt; + <version&gt;${oceanbase.version}</version&gt; + <scope&gt;provided</scope&gt; + </dependency&gt; It should not be imported directly due to the lgpl license https://github.com/oceanbase/obconnector-j In seatunnel-dist/pom.xml: &gt; @@ -488,6 +489,12 @@ </dependency&gt; <!-- jdbc driver --&gt; + <dependency&gt; + <groupId&gt;com.alipay.oceanbase</groupId&gt; + <artifactId&gt;oceanbase-client</artifactId&gt; + <version&gt;${oceanbase.version}</version&gt; + <scope&gt;provided</scope&gt; + </dependency&gt; Should not use the previous internal version, just keep it same with seatunnel-connectors-v2/connector-jdbc/pom.xml In seatunnel-e2e/seatunnel-connector-v2-e2e/connector-jdbc-e2e/connector-jdbc-e2e-part-2/src/test/java/org/apache/seatunnel/connectors/seatunnel/jdbc/JdbcOceanbaseIT.java: &gt; +import org.junit.jupiter.api.Disabled; +import org.testcontainers.containers.GenericContainer; +import org.testcontainers.containers.output.Slf4jLogConsumer; +import org.testcontainers.utility.DockerImageName; +import org.testcontainers.utility.DockerLoggerFactory; + +import com.google.common.collect.Lists; + +import java.util.ArrayList; +import java.util.HashMap; +import java.util.List; +import java.util.Map; + ***@***.***("Disabled because it needs user's personal oceanbase account to run this test!") +public class JdbcOceanbaseIT extends AbstractJdbcIT { + private static final String OCEANBASE_IMAGE = "shihd/oceanbase:1.0"; You can use the docker image oceanbase/oceanbase-ce:4.0.0.0 with user root and empty password. — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you authored the thread.Message ID: ***@***.***&gt;

Do you have time to complete this work recently? If not, we can also find other developers in the community to continue development based on your work.

**Pull request closed without merging**